realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -508,6 +508,45 @@ protected void checkIfValidAndInTransaction() {
         }
     }
 
+    /**
+     * Creates a row representing an embedded object - for internal use only.
+     *
+     * @param className the class name of the object to create.
+     * @param parentProxy The parent object which should hold a reference to the embedded object.
+     * @param parentProperty the property in the parent class which holds the reference.
+     * @param schema the Realm schema from which to obtain table information.
+     * @param parentObjectSchema the parent object schema from which to obtain property information.
+     * @return the row representing the newly created embedded object.
+     * @throws IllegalArgumentException if any embedded object invariants are broken.
+     */
+    Row getEmbeddedObjectRow(final String className,
+                             final RealmObjectProxy parentProxy,
+                             final String parentProperty,
+                             final RealmSchema schema,
+                             final RealmObjectSchema parentObjectSchema) {
+        final long parentPropertyColKey = parentObjectSchema.getColumnKey(parentProperty);
+        final RealmFieldType parentPropertyType = parentObjectSchema.getFieldType(parentProperty);
+        final String linkedType = parentObjectSchema.getLinkedType(parentProperty);
+        final Row row = parentProxy.realmGet$proxyState().getRow$realm();
+        Row embeddedObject;
+
+        switch (parentPropertyType) {
+            case OBJECT:
+            case LIST:
+                if (linkedType.equals(className)) {
+                    long objKey = row.createEmbeddedObject(parentPropertyColKey, parentPropertyType);
+                    embeddedObject = schema.getTable(className).getCheckedRow(objKey);
+                } else {
+                    throw new IllegalArgumentException(String.format("Parent type %s expects that property '%s' be of type %s but was %s.", parentObjectSchema.getClassName(), parentProperty, linkedType, className));
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("Parent property is not a reference to embedded objects of the appropriate type: " + parentPropertyType);
```

Comment:
I suspect we will never hit this as `parentObjectSchema.getLinkedType(parentProperty);` will throw first. We should probably move the `linkedType` property into the switch?

Reply:
That's right, and also having `parentProperty` as final makes the switch statement useless. Good catch.